{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8566db4e",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "deddab5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute protein descriptors\n",
    "from propy import PyPro\n",
    "from propy import AAComposition\n",
    "from propy import CTD\n",
    "\n",
    "# Build Sequence Object\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "\n",
    "# Read Fasta File\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "# Grouping iterable\n",
    "from itertools import chain\n",
    "\n",
    "# Return file path\n",
    "import glob\n",
    "\n",
    "# Unpack Files\n",
    "import json\n",
    "import pickle\n",
    "\n",
    "# Dataframes\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d29758a2",
   "metadata": {},
   "source": [
    "### Unpack all save models and features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6c8302ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files for Classification\n",
    "\n",
    "# Scaler\n",
    "file = open('Scaler Classification.pkl', 'rb')\n",
    "scaler_for_classification = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "# First Model\n",
    "file = open('SVM Linear Classification trained.pkl', 'rb')\n",
    "classification_model_1 = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "# Second Model\n",
    "file = open('SVM RBF Classification trained.pkl', 'rb')\n",
    "classification_model_2 = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "# Selected features Classification\n",
    "file = open(\"Features for Classification Model.json\")\n",
    "classification_features = json.loads(file.read())\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "266f3bc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Anaconda\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator MinMaxScaler from version 1.1.1 when using version 1.1.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "C:\\Anaconda\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator SVC from version 1.1.1 when using version 1.1.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "C:\\Anaconda\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator SVR from version 1.1.1 when using version 1.1.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Files for Regression\n",
    "\n",
    "# Scaler\n",
    "file = open('Scaler Regression.pickle', 'rb')\n",
    "scaler_for_regression = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "# Classification with ki\n",
    "file = open('SVC RBF bucket Classification trained.pickle', 'rb')\n",
    "classification_model_for_buckets = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "# Regression for Medium Bucket\n",
    "file = open('SVR RBF medium bucket Regression trained.pickle', 'rb')\n",
    "regression_model_medium_bucket = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "# Regression for Small Bucket\n",
    "file = open('SVR RBF small bucket Regression trained.pickle', 'rb')\n",
    "regression_model_small_bucket = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "# Selected features Regression\n",
    "file = open(\"Features for Regression Model.json\")\n",
    "regression_features = json.loads(file.read())\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cce51d71",
   "metadata": {},
   "source": [
    "### Inference Function"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "144df49b",
   "metadata": {},
   "source": [
    "#### Fasta File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "ae5dd4f8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def inferenceFasta(fastafile, ensemble=True):\n",
    "    \n",
    "    \"\"\" The inference function gets the protein sequence, trained model, preprocessing function and selected\n",
    "    features as input. \n",
    "\n",
    "    The function read the sequence as string and extract the peptide features using appropriate packages into \n",
    "    the dataframe.\n",
    "\n",
    "    The necessary features are selected from the extracted features which then undergoes preprocessing function, the\n",
    "    target value is predicted using trained function and give out the results. \"\"\"\n",
    "    \n",
    "    new_peptides = []\n",
    "    for file in glob.glob(fastafile):\n",
    "        new_peptides.append(file)\n",
    "        \n",
    "    for f in new_peptides:\n",
    "        fa = Fasta(f)\n",
    "        # empty list to save the features\n",
    "        allFeaturesData = []\n",
    "        for seq in fa:\n",
    "            # Make sure the sequence is a string\n",
    "            s = str(seq)\n",
    "            \n",
    "            # replace the unappropriate peptide sequence to A\n",
    "            s = s.replace('X','A')\n",
    "            s = s.replace('x','A')\n",
    "            s = s.replace('U','A')\n",
    "            s = s.replace('Z','A')\n",
    "            s = s.replace('B','A')\n",
    "\n",
    "            # Calculating primary features\n",
    "            analysed_seq = ProteinAnalysis(s)\n",
    "            wt = analysed_seq.molecular_weight()\n",
    "            arm = analysed_seq.aromaticity()\n",
    "            instab = analysed_seq.instability_index()\n",
    "            flex = analysed_seq.flexibility()\n",
    "            pI = analysed_seq.isoelectric_point()\n",
    "\n",
    "            # create a list for the primary features\n",
    "            pFeatures = [seq.name, s, len(seq), wt, arm, instab, pI]\n",
    "\n",
    "            # Get Amino Acid Composition (AAC), Composition Transition Distribution (CTD) and Dipeptide Composition (DPC)\n",
    "            resultAAC = AAComposition.CalculateAAComposition(s)\n",
    "            resultCTD = CTD.CalculateCTD(s)\n",
    "            resultDPC = AAComposition.CalculateDipeptideComposition(s)\n",
    "\n",
    "            # Collect all the features into lists\n",
    "            aacFeatures = [j for i,j in resultAAC.items()]\n",
    "            ctdFeatures = [l for k,l in resultCTD.items()]\n",
    "            dpcFeatures = [n for m,n in resultDPC.items()]\n",
    "            allFeaturesData.append(pFeatures + aacFeatures + ctdFeatures + dpcFeatures)\n",
    "        \n",
    "        # Collect feature names\n",
    "        pFeaturesName = ['Name','Seq' ,'SeqLength','Weight','Aromaticity','Instability','IsoelectricPoint']\n",
    "        aacFeaturesData = [i for i,j in resultAAC.items()]\n",
    "        ctdFeaturesData = [k for k,l in resultCTD.items()]\n",
    "        dpcFeaturesData = [m for m,n in resultDPC.items()]\n",
    "        \n",
    "        featuresName  = []\n",
    "        featuresName.append(pFeaturesName+aacFeaturesData+ctdFeaturesData+dpcFeaturesData)\n",
    "        featuresFlattenList = list(chain.from_iterable(featuresName))\n",
    "        \n",
    "        # create dataframe using all extracted features and the names\n",
    "        allFeaturesData = pd.DataFrame(allFeaturesData, columns = featuresFlattenList)\n",
    "        \n",
    "        \n",
    "        # CLASSIFICATION\n",
    "        # ----------------------------------\n",
    "        \n",
    "        # get only necessary features\n",
    "        selectedFeaturesClassification = allFeaturesData[classification_features]\n",
    "\n",
    "        # Apply preprocessing function\n",
    "        selectedFeaturesScaledClassification = pd.DataFrame(scaler_for_classification.transform(selectedFeaturesClassification),\n",
    "                                                            columns = selectedFeaturesClassification.columns)\n",
    "\n",
    "        # Model 1 - Applying threshold on decision function\n",
    "        decisionFuctionModel1 = classification_model_1.decision_function(selectedFeaturesScaledClassification)\n",
    "        threshold = 0.25\n",
    "        y_predict_model_1 = []\n",
    "        for j in decisionFuctionModel1:\n",
    "            if j > threshold:\n",
    "                y_predict_model_1.append(1)\n",
    "            else:\n",
    "                y_predict_model_1.append(0)\n",
    "                \n",
    "        # Model 2\n",
    "        y_predict_model_2 = classification_model_2.predict(selectedFeaturesScaledClassification)\n",
    "        \n",
    "        # Ensemble Model Prediction\n",
    "        if ensemble==True:\n",
    "            y_predict_ensemble = []\n",
    "            for i in range(len(y_predict_model_1)):\n",
    "                if (y_predict_model_1[i] == 1) & (y_predict_model_2[i] == 1):\n",
    "                    y_predict_ensemble.append('Positive')\n",
    "                else:\n",
    "                    y_predict_ensemble.append('Negative')\n",
    "            allFeaturesData = pd.concat([allFeaturesData, pd.DataFrame(y_predict_ensemble, columns=[\"Predicted\"])], axis=1)\n",
    "        \n",
    "        elif ensemble==False:\n",
    "            predictions_model_2 =  pd.DataFrame(y_predict_model_2, columns=[\"Predicted\"])\n",
    "            predictions_model_2.replace({1:'Positive',0:'Negative'}, inplace=True)\n",
    "            allFeaturesData = pd.concat([allFeaturesData, predictions_model_2], axis=1)\n",
    "        \n",
    "        # REGRESSION\n",
    "        # ----------------------------------\n",
    "        \n",
    "        # get positively predicted peptides\n",
    "        classifiedPeptides = allFeaturesData[allFeaturesData['Predicted']=='Positive']\n",
    "\n",
    "        # Apply preprocessing function and select only necessary features\n",
    "        selectedFeaturesScaledRegression = pd.DataFrame(scaler_for_regression.transform(classifiedPeptides.iloc[:,2:-1]), \n",
    "                                                            columns = classifiedPeptides.columns[2:-1])[regression_features]\n",
    "        \n",
    "        # Predict the buckets.\n",
    "        buckets_pred = classification_model_for_buckets.predict(selectedFeaturesScaledRegression)\n",
    "        selectedFeaturesScaledRegression['Bucket'] = buckets_pred\n",
    "        \n",
    "        # Fixed Ki range and Source Interval\n",
    "        ki_range = (-11.330603908176274, 17.19207365866807)\n",
    "        source_interval = (-5,5)\n",
    "        \n",
    "        # Make predictions for all of the buckets. The large bucket is predict as 0. Only make predictions if the arrays aren't empty.\n",
    "        if selectedFeaturesScaledRegression[buckets_pred==0].size != 0:\n",
    "            sml_pred = regression_model_small_bucket.predict(selectedFeaturesScaledRegression[buckets_pred==0].iloc[:,:-1])\n",
    "            sml_pred = np.exp(np.interp(sml_pred, source_interval, ki_range))\n",
    "        if selectedFeaturesScaledRegression[buckets_pred==1].size != 0:\n",
    "            med_pred = regression_model_medium_bucket.predict(selectedFeaturesScaledRegression[buckets_pred==1].iloc[:,:-1])\n",
    "            med_pred = np.exp(np.interp(med_pred, source_interval, ki_range))\n",
    "        lrg_pred = np.zeros(np.count_nonzero(selectedFeaturesScaledRegression[buckets_pred==2]))\n",
    "        \n",
    "        # Put back the predictions in the original order.\n",
    "        y_predict_regression = np.array([])\n",
    "        for i in buckets_pred:\n",
    "            if i == 0:\n",
    "                y_predict_regression = np.append(y_predict_regression, sml_pred[0])\n",
    "                sml_pred = np.delete(sml_pred, 0)\n",
    "            elif i == 1:\n",
    "                y_predict_regression = np.append(y_predict_regression, med_pred[0])\n",
    "                med_pred = np.delete(med_pred, 0)\n",
    "            elif i == 2:\n",
    "                y_predict_regression = np.append(y_predict_regression, lrg_pred[0])\n",
    "                lrg_pred = np.delete(lrg_pred, 0)\n",
    "        \n",
    "        classifiedPeptides['KI (nM) Predicted'] = y_predict_regression\n",
    "        \n",
    "        allFeaturesData = pd.merge(allFeaturesData,classifiedPeptides[['Seq','KI (nM) Predicted']],on='Seq', how='left')\n",
    "        \n",
    "        # save result in a new dataframe\n",
    "        result = allFeaturesData[['Name','Seq','Predicted','KI (nM) Predicted']]\n",
    "        \n",
    "        return result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3669402",
   "metadata": {},
   "source": [
    "#### Single Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "f4e08b01",
   "metadata": {},
   "outputs": [],
   "source": [
    "def inferenceSingleSeqence(seq, ensemble=True):\n",
    "    \n",
    "    \"\"\" The inference function gets the protein sequence, trained model, preprocessing function and selected\n",
    "    features as input. \n",
    "    \n",
    "    The function read the sequence as string and extract the peptide features using appropriate packages into \n",
    "    the dataframe.\n",
    "    \n",
    "    The necessary features are selected from the extracted features which then undergoes preprocessing function, the\n",
    "    target value is predicted using trained function and give out the results. \"\"\"\n",
    "    \n",
    "    # empty list to save the features\n",
    "    allFeaturesData = []\n",
    "    \n",
    "    # Make sure the sequence is a string\n",
    "    s = str(seq)\n",
    "    \n",
    "    # replace the unappropriate peptide sequence to A\n",
    "    s = s.replace('X','A')\n",
    "    s = s.replace('x','A')\n",
    "    s = s.replace('U','A')\n",
    "    s = s.replace('Z','A')\n",
    "    s = s.replace('B','A')\n",
    "    \n",
    "    # Calculating primary features\n",
    "    analysed_seq = ProteinAnalysis(s)\n",
    "    wt = analysed_seq.molecular_weight()\n",
    "    arm = analysed_seq.aromaticity()\n",
    "    instab = analysed_seq.instability_index()\n",
    "    flex = analysed_seq.flexibility()\n",
    "    pI = analysed_seq.isoelectric_point()\n",
    "    \n",
    "    # create a list for the primary features\n",
    "    pFeatures = [seq, s, len(seq), wt, arm, instab, pI]\n",
    "     \n",
    "    # Get Amino Acid Composition (AAC), Composition Transition Distribution (CTD) and Dipeptide Composition (DPC)\n",
    "    resultAAC = AAComposition.CalculateAAComposition(s)\n",
    "    resultCTD = CTD.CalculateCTD(s)\n",
    "    resultDPC = AAComposition.CalculateDipeptideComposition(s)\n",
    "    \n",
    "    # Collect all the features into lists\n",
    "    aacFeatures = [j for i,j in resultAAC.items()]\n",
    "    ctdFeatures = [l for k,l in resultCTD.items()]\n",
    "    dpcFeatures = [n for m,n in resultDPC.items()]\n",
    "    allFeaturesData.append(pFeatures + aacFeatures + ctdFeatures + dpcFeatures)\n",
    "    \n",
    "    # Collect feature names\n",
    "    name1 = ['Name','Seq' ,'SeqLength','Weight','Aromaticity','Instability','IsoelectricPoint']\n",
    "    name2 = [i for i,j in resultAAC.items()]\n",
    "    name3 = [k for k,l in resultCTD.items()]\n",
    "    name4 = [m for m,n in resultDPC.items()]\n",
    "    name  = []\n",
    "    name.append(name1+name2+name3+name4)\n",
    "    flatten_list = list(chain.from_iterable(name))\n",
    "    \n",
    "    # create dataframe using all extracted features and the names\n",
    "    allFeaturesData = pd.DataFrame(allFeaturesData, columns = flatten_list)\n",
    "        \n",
    "        \n",
    "    # CLASSIFICATION\n",
    "    # ----------------------------------\n",
    "\n",
    "    # get only necessary features\n",
    "    selectedFeaturesClassification = allFeaturesData[classification_features]\n",
    "\n",
    "    # Apply preprocessing function\n",
    "    selectedFeaturesScaledClassification = pd.DataFrame(scaler_for_classification.transform(selectedFeaturesClassification),\n",
    "                                                        columns = selectedFeaturesClassification.columns)\n",
    "\n",
    "    # Model 1 - Applying threshold on decision function\n",
    "    decisionFuctionModel1 = classification_model_1.decision_function(selectedFeaturesScaledClassification)\n",
    "    threshold = 0.25\n",
    "    y_predict_model_1 = []\n",
    "    for j in decisionFuctionModel1:\n",
    "        if j > threshold:\n",
    "            y_predict_model_1.append(1)\n",
    "        else:\n",
    "            y_predict_model_1.append(0)\n",
    "\n",
    "    # Model 2\n",
    "    y_predict_model_2 = classification_model_2.predict(selectedFeaturesScaledClassification)\n",
    "\n",
    "    # Ensemble Model Prediction\n",
    "    if ensemble==True:\n",
    "        y_predict_ensemble = []\n",
    "        for i in range(len(y_predict_model_1)):\n",
    "            if (y_predict_model_1[i] == 1) & (y_predict_model_2[i] == 1):\n",
    "                y_predict_ensemble.append('Positive')\n",
    "            else:\n",
    "                y_predict_ensemble.append('Negative')\n",
    "        allFeaturesData = pd.concat([allFeaturesData, pd.DataFrame(y_predict_ensemble, columns=[\"Predicted\"])], axis=1)\n",
    "\n",
    "    elif ensemble==False:\n",
    "        predictions_model_2 =  pd.DataFrame(y_predict_model_2, columns=[\"Predicted\"])\n",
    "        predictions_model_2.replace({1:'Positive',0:'Negative'}, inplace=True)\n",
    "        allFeaturesData = pd.concat([allFeaturesData, predictions_model_2], axis=1)\n",
    "    \n",
    "    # REGRESSION\n",
    "    # ----------------------------------\n",
    "\n",
    "    # get positively predicted peptides\n",
    "    classifiedPeptides = allFeaturesData[allFeaturesData['Predicted']=='Positive']\n",
    "    \n",
    "    # Exception for negative peptides\n",
    "    if len(classifiedPeptides) == 0:\n",
    "         # save result in a new dataframe\n",
    "        result = allFeaturesData[['Name','Seq','Predicted']]\n",
    "        return result\n",
    "        \n",
    "\n",
    "    # Apply preprocessing function and select only necessary features\n",
    "    selectedFeaturesScaledRegression = pd.DataFrame(scaler_for_regression.transform(classifiedPeptides.iloc[:,2:-1]), \n",
    "                                                        columns = classifiedPeptides.columns[2:-1])[regression_features]\n",
    "\n",
    "    # Predict the buckets.\n",
    "    buckets_pred = classification_model_for_buckets.predict(selectedFeaturesScaledRegression)\n",
    "    selectedFeaturesScaledRegression['Bucket'] = buckets_pred\n",
    "\n",
    "    # Fixed Ki range and Source Interval\n",
    "    ki_range = (-11.330603908176274, 17.19207365866807)\n",
    "    source_interval = (-5,5)\n",
    "\n",
    "    # Make predictions for all of the buckets. The large bucket is predict as 0. Only make predictions if the arrays aren't empty.\n",
    "    if selectedFeaturesScaledRegression[buckets_pred==0].size != 0:\n",
    "        sml_pred = regression_model_small_bucket.predict(selectedFeaturesScaledRegression[buckets_pred==0].iloc[:,:-1])\n",
    "        sml_pred = np.exp(np.interp(sml_pred, source_interval, ki_range))\n",
    "    if selectedFeaturesScaledRegression[buckets_pred==1].size != 0:\n",
    "        med_pred = regression_model_medium_bucket.predict(selectedFeaturesScaledRegression[buckets_pred==1].iloc[:,:-1])\n",
    "        med_pred = np.exp(np.interp(med_pred, source_interval, ki_range))\n",
    "    lrg_pred = np.zeros(np.count_nonzero(selectedFeaturesScaledRegression[buckets_pred==2]))\n",
    "\n",
    "    # Put back the predictions in the original order.\n",
    "    y_predict_regression = np.array([])\n",
    "    for i in buckets_pred:\n",
    "        if i == 0:\n",
    "            y_predict_regression = np.append(y_predict_regression, sml_pred[0])\n",
    "            sml_pred = np.delete(sml_pred, 0)\n",
    "        elif i == 1:\n",
    "            y_predict_regression = np.append(y_predict_regression, med_pred[0])\n",
    "            med_pred = np.delete(med_pred, 0)\n",
    "        elif i == 2:\n",
    "            y_predict_regression = np.append(y_predict_regression, lrg_pred[0])\n",
    "            lrg_pred = np.delete(lrg_pred, 0)\n",
    "\n",
    "    classifiedPeptides['KI (nM) Predicted'] = y_predict_regression\n",
    "\n",
    "    allFeaturesData = pd.merge(allFeaturesData,classifiedPeptides[['Seq','KI (nM) Predicted']],on='Seq', how='left')\n",
    "\n",
    "    # save result in a new dataframe\n",
    "    result = allFeaturesData[['Name','Seq','Predicted','KI (nM) Predicted']]\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "29246010",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Predicted</th>\n",
       "      <th>KI (nM) Predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APEADQTTPEEKPAEPEPVA</td>\n",
       "      <td>APEADQTTPEEKPAEPEPVA</td>\n",
       "      <td>Positive</td>\n",
       "      <td>1.663297</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Name                   Seq Predicted  KI (nM) Predicted\n",
       "0  APEADQTTPEEKPAEPEPVA  APEADQTTPEEKPAEPEPVA  Positive           1.663297"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inferenceSingleSeqence('APEADQTTPEEKPAEPEPVA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "760ec0d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Predicted</th>\n",
       "      <th>KI (nM) Predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>QSPLPERQE</td>\n",
       "      <td>QSPLPERQE</td>\n",
       "      <td>Positive</td>\n",
       "      <td>1.207007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Name        Seq Predicted  KI (nM) Predicted\n",
       "0  QSPLPERQE  QSPLPERQE  Positive           1.207007"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inferenceSingleSeqence('QSPLPERQE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "fc9e4c2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HTLGYINDNEEGPR</td>\n",
       "      <td>HTLGYINDNEEGPR</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Name             Seq Predicted\n",
       "0  HTLGYINDNEEGPR  HTLGYINDNEEGPR  Negative"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inferenceSingleSeqence('HTLGYINDNEEGPR')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c8c4f90",
   "metadata": {},
   "source": [
    "### Testing the Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "f4e01b2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-181-1d033c624145>:145: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  classifiedPeptides['KI (nM) Predicted'] = y_predict_regression\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Predicted</th>\n",
       "      <th>KI (nM) Predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P80849</td>\n",
       "      <td>EDLPEK</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q7M066</td>\n",
       "      <td>FEQNTAQA</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A5K0XKV1</td>\n",
       "      <td>EKFEGPGVK</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A6A3HBE0</td>\n",
       "      <td>QSPLPERQE</td>\n",
       "      <td>Positive</td>\n",
       "      <td>1.207007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A5K1D874</td>\n",
       "      <td>DHLNAEQGK</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q2UVK8</td>\n",
       "      <td>EGISFPKFEN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A5B7JC83</td>\n",
       "      <td>MERQGSGREE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A0A5K1ESP7</td>\n",
       "      <td>VEKFYQQCDP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A022PZR9</td>\n",
       "      <td>SYQCRPFQQL</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>D9U971</td>\n",
       "      <td>GTEGCENAKP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A0A5K1EJS8</td>\n",
       "      <td>GLISGNQAEN</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A0A5K1DW74</td>\n",
       "      <td>TESGEPIDQD</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A0A6V7HZ22</td>\n",
       "      <td>VQVTDESGEQ</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A0A5K0XWV6</td>\n",
       "      <td>DAQANGKPLQ</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Q3UTH2</td>\n",
       "      <td>MEKTLETVPLE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A6V7JP58</td>\n",
       "      <td>MASAMTRSWKL</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>D2SXU1</td>\n",
       "      <td>LPWGQMSFWGA</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A0A6V7M665</td>\n",
       "      <td>LYPRVQPSKSD</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A0A4P9D3R7</td>\n",
       "      <td>MSPQTETKASV</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A0A6V7I3H9</td>\n",
       "      <td>MKMEEDREFWEV</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A0A5K0YWZ1</td>\n",
       "      <td>SGLPSNANTTVA</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Q9PVC2</td>\n",
       "      <td>CFKKDMHKVETY</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A0A5K0Z870</td>\n",
       "      <td>FLFLMARSIEEP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A0A7R5U2U5</td>\n",
       "      <td>MEEFQVYLELDR</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A0A126TR29</td>\n",
       "      <td>NLAPSFVNEDGVE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>K4IU71</td>\n",
       "      <td>CFKKDMHKVETY</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A0A5K1ERB3</td>\n",
       "      <td>WIGRLVLPKKCVE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A0A4Y2UTJ5</td>\n",
       "      <td>LDLKITYLNSYVF</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A0A5K0YGQ4</td>\n",
       "      <td>NAREDSTKQMLTE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A0A7R8ZV56</td>\n",
       "      <td>MCNPPELKPSKTS</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>E4YZ42</td>\n",
       "      <td>LRNDESLESSSDG</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A0A5K0WMU8</td>\n",
       "      <td>MVELKKQENIDTT</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>A0A5K0V6Q3</td>\n",
       "      <td>HTLGYINDNEEGPR</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>A0A7U0QQG7</td>\n",
       "      <td>YADIGGLDNQIQEIK</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>A0A141GB61</td>\n",
       "      <td>MIIRSPEPEVKIVVD</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A0A5K1GA59</td>\n",
       "      <td>KLDEPLSRIATYLTG</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>C0KC49</td>\n",
       "      <td>FPLDLASVEAPSING</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>A0A5K0VQN6</td>\n",
       "      <td>AERLQNLMEQDYDIG</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>A0A5K0VT39</td>\n",
       "      <td>QQPEFHMTMISTLLR</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>A0A6V7HRJ8</td>\n",
       "      <td>MTILICSLLVYGAWK</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Q6MZZ1</td>\n",
       "      <td>MASLLKVDQEVKLKL</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>AIP27</td>\n",
       "      <td>APAEPEAPAPETEKAEEVEK</td>\n",
       "      <td>Positive</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>AIP28</td>\n",
       "      <td>APEADQTTPEEKPAEPEPVA</td>\n",
       "      <td>Positive</td>\n",
       "      <td>1.663297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>AIP380</td>\n",
       "      <td>EKFPSSPPTTPPSPAKTD</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>AIP1092</td>\n",
       "      <td>VPRLPEQGSSSRAEDSPEG</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Fengycin</td>\n",
       "      <td>EKYTEVPEYV</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Name                   Seq Predicted  KI (nM) Predicted\n",
       "0       P80849                EDLPEK  Negative                NaN\n",
       "1       Q7M066              FEQNTAQA  Negative                NaN\n",
       "2   A0A5K0XKV1             EKFEGPGVK  Negative                NaN\n",
       "3   A0A6A3HBE0             QSPLPERQE  Positive           1.207007\n",
       "4   A0A5K1D874             DHLNAEQGK  Negative                NaN\n",
       "5       Q2UVK8            EGISFPKFEN  Negative                NaN\n",
       "6   A0A5B7JC83            MERQGSGREE  Negative                NaN\n",
       "7   A0A5K1ESP7            VEKFYQQCDP  Negative                NaN\n",
       "8   A0A022PZR9            SYQCRPFQQL  Negative                NaN\n",
       "9       D9U971            GTEGCENAKP  Negative                NaN\n",
       "10  A0A5K1EJS8            GLISGNQAEN  Negative                NaN\n",
       "11  A0A5K1DW74            TESGEPIDQD  Negative                NaN\n",
       "12  A0A6V7HZ22            VQVTDESGEQ  Negative                NaN\n",
       "13  A0A5K0XWV6            DAQANGKPLQ  Negative                NaN\n",
       "14      Q3UTH2           MEKTLETVPLE  Negative                NaN\n",
       "15  A0A6V7JP58           MASAMTRSWKL  Negative                NaN\n",
       "16      D2SXU1           LPWGQMSFWGA  Negative                NaN\n",
       "17  A0A6V7M665           LYPRVQPSKSD  Negative                NaN\n",
       "18  A0A4P9D3R7           MSPQTETKASV  Negative                NaN\n",
       "19  A0A6V7I3H9          MKMEEDREFWEV  Negative                NaN\n",
       "20  A0A5K0YWZ1          SGLPSNANTTVA  Negative                NaN\n",
       "21      Q9PVC2          CFKKDMHKVETY  Negative                NaN\n",
       "22  A0A5K0Z870          FLFLMARSIEEP  Negative                NaN\n",
       "23  A0A7R5U2U5          MEEFQVYLELDR  Negative                NaN\n",
       "24  A0A126TR29         NLAPSFVNEDGVE  Negative                NaN\n",
       "25      K4IU71          CFKKDMHKVETY  Negative                NaN\n",
       "26  A0A5K1ERB3         WIGRLVLPKKCVE  Negative                NaN\n",
       "27  A0A4Y2UTJ5         LDLKITYLNSYVF  Negative                NaN\n",
       "28  A0A5K0YGQ4         NAREDSTKQMLTE  Negative                NaN\n",
       "29  A0A7R8ZV56         MCNPPELKPSKTS  Negative                NaN\n",
       "30      E4YZ42         LRNDESLESSSDG  Negative                NaN\n",
       "31  A0A5K0WMU8         MVELKKQENIDTT  Negative                NaN\n",
       "32  A0A5K0V6Q3        HTLGYINDNEEGPR  Negative                NaN\n",
       "33  A0A7U0QQG7       YADIGGLDNQIQEIK  Negative                NaN\n",
       "34  A0A141GB61       MIIRSPEPEVKIVVD  Negative                NaN\n",
       "35  A0A5K1GA59       KLDEPLSRIATYLTG  Negative                NaN\n",
       "36      C0KC49       FPLDLASVEAPSING  Negative                NaN\n",
       "37  A0A5K0VQN6       AERLQNLMEQDYDIG  Negative                NaN\n",
       "38  A0A5K0VT39       QQPEFHMTMISTLLR  Negative                NaN\n",
       "39  A0A6V7HRJ8       MTILICSLLVYGAWK  Negative                NaN\n",
       "40      Q6MZZ1       MASLLKVDQEVKLKL  Negative                NaN\n",
       "41       AIP27  APAEPEAPAPETEKAEEVEK  Positive           0.000000\n",
       "42       AIP28  APEADQTTPEEKPAEPEPVA  Positive           1.663297\n",
       "43      AIP380    EKFPSSPPTTPPSPAKTD  Negative                NaN\n",
       "44     AIP1092   VPRLPEQGSSSRAEDSPEG  Negative                NaN\n",
       "45    Fengycin            EKYTEVPEYV  Negative                NaN"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inferenceFasta('smallhits.fasta', ensemble=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "1ee85629",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-181-1d033c624145>:145: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  classifiedPeptides['KI (nM) Predicted'] = y_predict_regression\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Predicted</th>\n",
       "      <th>KI (nM) Predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A699VUS0</td>\n",
       "      <td>THDDVDQENVVEETVDDVAQPTSPLPPSPSVPPSPPHQSPRSSPSQ...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A7I9YXR2</td>\n",
       "      <td>MSAQDKVKNKIEDVSGKAKEALGKATNDPGVRDEGRGDQTKASLKD...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0A6ZHE3</td>\n",
       "      <td>MVQIKFLFAFLAVMTIVVLAANMADADFLSGKFKGGCMMWSTEKCR...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A7W8YT97</td>\n",
       "      <td>MFKQFLDKVDGNQGYLLSSLGIFMLFFLLVGILLLTMKKDDIKYMS...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1P496</td>\n",
       "      <td>MHGSICSSANYLPLPTLPALGRRGGSLVHGGFSVSETSFPPLRLGA...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q5Z0N8</td>\n",
       "      <td>MPVSATESSTTPTARPAIEGESPVVWPDPSPLTGWWERVMRGGPVD...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A4V2REQ9</td>\n",
       "      <td>MGAKFEVFKDGRGEYRFRLKAPNGQIIASSEGYKSKDSALNGVASV...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A0A375IAT7</td>\n",
       "      <td>MKKLIAALVVGLFATGAFAQAAAPAPAEPAAPAATKEAPKKTTKKK...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A524LF58</td>\n",
       "      <td>MSGQGSTGNVIAALASFFIPGLGQLLQGRLLIAIVMFVLAAALWFI...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A0A7Z0HL73</td>\n",
       "      <td>MTTPELTRREKRPAERYGERTARPEKLRNIEVWARSAPIRLAGYED...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A0A0L1NF93</td>\n",
       "      <td>MNIQRAKELSVSAEHANVSFQGMPVMIQHVDESNETARIYEVKNPG...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>H1QGQ7</td>\n",
       "      <td>MHRWRFESADGDDGRPAGDGEGRPAGGHEGRSAEGDGDQQARGEAD...</td>\n",
       "      <td>Positive</td>\n",
       "      <td>3.953144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>H4IHK7</td>\n",
       "      <td>MNNNEPDTLPDPAIGYIFQNDILALKQAFSLPDIDYADISQREQLA...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A0A2U8GW77</td>\n",
       "      <td>MPGIRVKENEPFEVAIRRFKRTIEKTGVLTELRSREFYEKPTAERK...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A0A7K2P8R6</td>\n",
       "      <td>ATGTPSGAVPEVGAVEQPVVEIDVEAPAVAVQGAYDTAGPRQAAAT...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A6B5IMJ6</td>\n",
       "      <td>EDGDSDLTIEEIDLIGSHLDKEIEDCAQIRKQTRKKELKDETVDAH...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A0A3B9LF71</td>\n",
       "      <td>MRNGVQFLTDETGKKTGVVLPVSEYEKLLEDLQDLAVVAERRGEPS...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>B4Y7L8</td>\n",
       "      <td>TDSSIISAGWEPTEENEDERDRHPSYSGSGIDDDEDFI</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>G4HQW5</td>\n",
       "      <td>MRIRSNPAKCVGAGTCAALAPAVFDQDENSGIVEVVDENPPESERK...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A0A1H6WJY6</td>\n",
       "      <td>MKNDIAIGQMQGLRWASNEARRIASLYPPSSDGQNTFALHADACDR...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>E8JVV0</td>\n",
       "      <td>MAQIKITLTKSPIGRIPSQRKTVVALGLGKLNSSVIKEDNPAVRGM...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>R5Q569</td>\n",
       "      <td>MVKIENYLKENGESKTNDIAEYLNLSSARTRKILSEMKTIEAIGTN...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A0A7W7AJZ1</td>\n",
       "      <td>MSKIDEVQSEIKATPGLGKKMAKYGAIGAVVAIPIPFVGPVFGALA...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>W4JSH1</td>\n",
       "      <td>RKTAQKSPLQPSGNLYLPALPQQQKAREANTKWQTILRQVRTTNLT...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A0A7X2IYC3</td>\n",
       "      <td>MMAQIRLLGKREEVNKIIQSFEQDFTITSTSREVCNLKDREEGQKH...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A0A6J7GK72</td>\n",
       "      <td>MRWTDAVNALYRIELHRNPTALAANGGAWRGLDDPTPAKAEAEAEA...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A0A7W6VXK9</td>\n",
       "      <td>MLEGAAGVFGADASTAQGPPVDLKALHAKIGELALENDFLEGALIK...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A0A7W1WS82</td>\n",
       "      <td>MKELIIKAKGKRGRDIEVLVDGQPVEGLKYIQFEADIEGPIELSVM...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A0A7X0G9P5</td>\n",
       "      <td>MALWPLTTVGAPWVPLAAAVPTAAALAIILASRRGGYPAGPRETRS...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A0A1B2IVD2</td>\n",
       "      <td>MENRVNESRQNLQQCEFATQSERLTQRVAKLLHLQPIETATIHGGH...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A0A1I5X0U8</td>\n",
       "      <td>MAFDVSYLLLQVTQQCNLCCPYCFYSEEYPNTRSATNKKMNIETAK...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A0A538MMP2</td>\n",
       "      <td>TLVRTLLEHDLVDELRLMVYPILLGSGKRLFGDSASGKAFRLLELR...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>A0A434JR87</td>\n",
       "      <td>IEDGEKPAGSGACLGHIPLAEGIERFGA</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>A0A0K2T0Y8</td>\n",
       "      <td>MFNESPILTLRACKSPITQFVLLFETFLFLKELSDSVPLRSLFLIF...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>A0A150KLU2</td>\n",
       "      <td>MQIEIEIKKMRVITDSDATSEQLLQALNKTLLAGVEPLEIIQKLYL...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A0A6A6DU41</td>\n",
       "      <td>MLAVLEPTLHTLIFLIYPVNAKTLVLLVARHCRLIVISRLSDTKLN...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>A0A7X2WR87</td>\n",
       "      <td>MDRKISYPHFGDNSGDNPLSYPHTGHNLVDSLGITFRAHNRCLRNA...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>A0A1G7K008</td>\n",
       "      <td>MKLTKLKLARLRAGLKQTELAKMLGVSPSYINKIENNAVKPTMELL...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>A0A7K1SF48</td>\n",
       "      <td>MCHTPEVRGYGDKKPGGLASEPANVARPEFWWAGKAAGLGNPPHRR...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>A0A3M7PSL2</td>\n",
       "      <td>NFSGKILFCCFINNVKRNTLAIEKVSYETPKPSNEAVQSQSIQVVF...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>A0A7K1YEH0</td>\n",
       "      <td>MEKFSKLKELIASVEADAEKFYNNSNNAAGTRVRKAMQDLKNLAQE...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>S9VW97</td>\n",
       "      <td>MGGYSVLLVGLLTALLCVISYYFAPKGSNTNTWQMTLILTFSCCYI...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>M6IAV0</td>\n",
       "      <td>MTLVLTKASSVKKFSTNFTKLGLKSTSSFIDFDSFCTGSEIFSFEI...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>A0A392PQT1</td>\n",
       "      <td>MFERSSVSPPAGISLPPLSSLNHHE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>A0A173J4X2</td>\n",
       "      <td>MPTQNPHRIVGLCTSSKVYNALTELKHLEGHRTAKFLSLLAENLVR...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>T0N6R1</td>\n",
       "      <td>MSSFWLKVLACVIMLIDHVGAIFFPKLWGLRVIGRLAFPIFSFFIS...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>A0A7V7JLU7</td>\n",
       "      <td>AEELSPPASVATVGGLLDWLKTRSERHAAAFAEPSRLRAAVNQDFA...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>A0A6M6ICZ3</td>\n",
       "      <td>MFRARQILVRTIVLAVSTAAFAACGQRGPLYLPTDPAAAQRATLPE...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>T4VPV3</td>\n",
       "      <td>MNKQTICKDKNCKFFKTCLIKDVEILKKCTGREVKTSNQKVVG</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>A0A1X3MW98</td>\n",
       "      <td>MLCTFNVNIKHSNLHNKNIDKHIVLKFLGTAVMGILESYVLGEIDN...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>A0A5Q3QS16</td>\n",
       "      <td>MNMNTKDHCYAKSIILQAVYFKLRFTLSYRDVEEIIKMGVVFVDHA...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>V6UN32</td>\n",
       "      <td>MAGDFLGQGPLLLAAAALGPLSVPARSTGATARADWALSADVLLVG...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>A0A395Z784</td>\n",
       "      <td>MSIMHKKLMQIVLVKGLDIQDIVTLKKYDKKVKKFLTNKIGDDILI...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>A0A349XYA6</td>\n",
       "      <td>MSGFTTVLALIVMGLVVIVLIRGLFNMLKGQDANRSNKLMQLRLLL...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>A0A1M6C5T9</td>\n",
       "      <td>MILRGIALSLGKVLAVLAALAALASRFTSCLLCVYQPEMPQKVRNM...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>A0A6I5N5G3</td>\n",
       "      <td>MEDFFRNVSRYPRYFISFTLGVLYNALQPLLPLLKRPTTAVALIGS...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>A0A1H7WRT9</td>\n",
       "      <td>MKKLSNLEGVVVLSKKAQKTVHGGDYVMSSCSDYCPGTVQGNIFTG...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>A0A7W0SMF1</td>\n",
       "      <td>MTDPRNDDKKDGSNWVGGSVLGVGIGAGVGAIVGNLGNGIWIGAVI...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>B4Y7M7</td>\n",
       "      <td>TDSSIISAGWEPTEENEDERDRHPSYSGSGIDDDEDFI</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>A0A1C4D5Q1</td>\n",
       "      <td>MGRTKQGNRNAQANTNAKRGNRTSSEMVEFTTGRAKAEKNRPQD</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>A0A2H6JUL2</td>\n",
       "      <td>MLIGEAEFKLRAVKILFENGEYSDVISRALSYLVIS</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>A0A379M3S6</td>\n",
       "      <td>MGLRGSGSAARTHPLSAAPNFFTSGLHAAFDDGMQAVEWNFAGQGP...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>B0N079</td>\n",
       "      <td>MKRIYRSRQDRMVCGVCGGIAEYFDLDPSLVRLGWIIFSAMGGSGF...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>A0A1M6PH39</td>\n",
       "      <td>MNLDTFQNKLILILSYVDKLKRENVPINTQRILIQTYANDLEINLT...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>A0A0A0NJ57</td>\n",
       "      <td>MREPCISDRDRGWFKSSYSGAGNTECVEAAFRPGSTAVRDSKDPGG...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>A0A7Z2XC48</td>\n",
       "      <td>MNLSSMVKGILLALIFLGLMAGVLPSGAQLVRMGGSSWFLVGQSAE...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>A0A1V3WIU2</td>\n",
       "      <td>MHRPIQPDSQISTGLTSRSGNAGDTAADAELRIGPQRSAHRLSAVV...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>A0A1M6SHT4</td>\n",
       "      <td>MKRTFQPHNRKRVNKHGFMTRMEDRWGRAVLSRRRAKGRKRLTISD...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Name                                                Seq Predicted  \\\n",
       "0   A0A699VUS0  THDDVDQENVVEETVDDVAQPTSPLPPSPSVPPSPPHQSPRSSPSQ...  Negative   \n",
       "1   A0A7I9YXR2  MSAQDKVKNKIEDVSGKAKEALGKATNDPGVRDEGRGDQTKASLKD...  Negative   \n",
       "2   A0A0A6ZHE3  MVQIKFLFAFLAVMTIVVLAANMADADFLSGKFKGGCMMWSTEKCR...  Negative   \n",
       "3   A0A7W8YT97  MFKQFLDKVDGNQGYLLSSLGIFMLFFLLVGILLLTMKKDDIKYMS...  Negative   \n",
       "4       L1P496  MHGSICSSANYLPLPTLPALGRRGGSLVHGGFSVSETSFPPLRLGA...  Negative   \n",
       "5       Q5Z0N8  MPVSATESSTTPTARPAIEGESPVVWPDPSPLTGWWERVMRGGPVD...  Negative   \n",
       "6   A0A4V2REQ9  MGAKFEVFKDGRGEYRFRLKAPNGQIIASSEGYKSKDSALNGVASV...  Negative   \n",
       "7   A0A375IAT7  MKKLIAALVVGLFATGAFAQAAAPAPAEPAAPAATKEAPKKTTKKK...  Negative   \n",
       "8   A0A524LF58  MSGQGSTGNVIAALASFFIPGLGQLLQGRLLIAIVMFVLAAALWFI...  Negative   \n",
       "9   A0A7Z0HL73  MTTPELTRREKRPAERYGERTARPEKLRNIEVWARSAPIRLAGYED...  Negative   \n",
       "10  A0A0L1NF93  MNIQRAKELSVSAEHANVSFQGMPVMIQHVDESNETARIYEVKNPG...  Negative   \n",
       "11      H1QGQ7  MHRWRFESADGDDGRPAGDGEGRPAGGHEGRSAEGDGDQQARGEAD...  Positive   \n",
       "12      H4IHK7  MNNNEPDTLPDPAIGYIFQNDILALKQAFSLPDIDYADISQREQLA...  Negative   \n",
       "13  A0A2U8GW77  MPGIRVKENEPFEVAIRRFKRTIEKTGVLTELRSREFYEKPTAERK...  Negative   \n",
       "14  A0A7K2P8R6  ATGTPSGAVPEVGAVEQPVVEIDVEAPAVAVQGAYDTAGPRQAAAT...  Negative   \n",
       "15  A0A6B5IMJ6  EDGDSDLTIEEIDLIGSHLDKEIEDCAQIRKQTRKKELKDETVDAH...  Negative   \n",
       "16  A0A3B9LF71  MRNGVQFLTDETGKKTGVVLPVSEYEKLLEDLQDLAVVAERRGEPS...  Negative   \n",
       "17      B4Y7L8             TDSSIISAGWEPTEENEDERDRHPSYSGSGIDDDEDFI  Negative   \n",
       "18      G4HQW5  MRIRSNPAKCVGAGTCAALAPAVFDQDENSGIVEVVDENPPESERK...  Negative   \n",
       "19  A0A1H6WJY6  MKNDIAIGQMQGLRWASNEARRIASLYPPSSDGQNTFALHADACDR...  Negative   \n",
       "20      E8JVV0  MAQIKITLTKSPIGRIPSQRKTVVALGLGKLNSSVIKEDNPAVRGM...  Negative   \n",
       "21      R5Q569  MVKIENYLKENGESKTNDIAEYLNLSSARTRKILSEMKTIEAIGTN...  Negative   \n",
       "22  A0A7W7AJZ1  MSKIDEVQSEIKATPGLGKKMAKYGAIGAVVAIPIPFVGPVFGALA...  Negative   \n",
       "23      W4JSH1  RKTAQKSPLQPSGNLYLPALPQQQKAREANTKWQTILRQVRTTNLT...  Negative   \n",
       "24  A0A7X2IYC3  MMAQIRLLGKREEVNKIIQSFEQDFTITSTSREVCNLKDREEGQKH...  Negative   \n",
       "25  A0A6J7GK72  MRWTDAVNALYRIELHRNPTALAANGGAWRGLDDPTPAKAEAEAEA...  Negative   \n",
       "26  A0A7W6VXK9  MLEGAAGVFGADASTAQGPPVDLKALHAKIGELALENDFLEGALIK...  Negative   \n",
       "27  A0A7W1WS82  MKELIIKAKGKRGRDIEVLVDGQPVEGLKYIQFEADIEGPIELSVM...  Negative   \n",
       "28  A0A7X0G9P5  MALWPLTTVGAPWVPLAAAVPTAAALAIILASRRGGYPAGPRETRS...  Negative   \n",
       "29  A0A1B2IVD2  MENRVNESRQNLQQCEFATQSERLTQRVAKLLHLQPIETATIHGGH...  Negative   \n",
       "30  A0A1I5X0U8  MAFDVSYLLLQVTQQCNLCCPYCFYSEEYPNTRSATNKKMNIETAK...  Negative   \n",
       "31  A0A538MMP2  TLVRTLLEHDLVDELRLMVYPILLGSGKRLFGDSASGKAFRLLELR...  Negative   \n",
       "32  A0A434JR87                       IEDGEKPAGSGACLGHIPLAEGIERFGA  Negative   \n",
       "33  A0A0K2T0Y8  MFNESPILTLRACKSPITQFVLLFETFLFLKELSDSVPLRSLFLIF...  Negative   \n",
       "34  A0A150KLU2  MQIEIEIKKMRVITDSDATSEQLLQALNKTLLAGVEPLEIIQKLYL...  Negative   \n",
       "35  A0A6A6DU41  MLAVLEPTLHTLIFLIYPVNAKTLVLLVARHCRLIVISRLSDTKLN...  Negative   \n",
       "36  A0A7X2WR87  MDRKISYPHFGDNSGDNPLSYPHTGHNLVDSLGITFRAHNRCLRNA...  Negative   \n",
       "37  A0A1G7K008  MKLTKLKLARLRAGLKQTELAKMLGVSPSYINKIENNAVKPTMELL...  Negative   \n",
       "38  A0A7K1SF48  MCHTPEVRGYGDKKPGGLASEPANVARPEFWWAGKAAGLGNPPHRR...  Negative   \n",
       "39  A0A3M7PSL2  NFSGKILFCCFINNVKRNTLAIEKVSYETPKPSNEAVQSQSIQVVF...  Negative   \n",
       "40  A0A7K1YEH0  MEKFSKLKELIASVEADAEKFYNNSNNAAGTRVRKAMQDLKNLAQE...  Negative   \n",
       "41      S9VW97  MGGYSVLLVGLLTALLCVISYYFAPKGSNTNTWQMTLILTFSCCYI...  Negative   \n",
       "42      M6IAV0  MTLVLTKASSVKKFSTNFTKLGLKSTSSFIDFDSFCTGSEIFSFEI...  Negative   \n",
       "43  A0A392PQT1                          MFERSSVSPPAGISLPPLSSLNHHE  Negative   \n",
       "44  A0A173J4X2  MPTQNPHRIVGLCTSSKVYNALTELKHLEGHRTAKFLSLLAENLVR...  Negative   \n",
       "45      T0N6R1  MSSFWLKVLACVIMLIDHVGAIFFPKLWGLRVIGRLAFPIFSFFIS...  Negative   \n",
       "46  A0A7V7JLU7  AEELSPPASVATVGGLLDWLKTRSERHAAAFAEPSRLRAAVNQDFA...  Negative   \n",
       "47  A0A6M6ICZ3  MFRARQILVRTIVLAVSTAAFAACGQRGPLYLPTDPAAAQRATLPE...  Negative   \n",
       "48      T4VPV3        MNKQTICKDKNCKFFKTCLIKDVEILKKCTGREVKTSNQKVVG  Negative   \n",
       "49  A0A1X3MW98  MLCTFNVNIKHSNLHNKNIDKHIVLKFLGTAVMGILESYVLGEIDN...  Negative   \n",
       "50  A0A5Q3QS16  MNMNTKDHCYAKSIILQAVYFKLRFTLSYRDVEEIIKMGVVFVDHA...  Negative   \n",
       "51      V6UN32  MAGDFLGQGPLLLAAAALGPLSVPARSTGATARADWALSADVLLVG...  Negative   \n",
       "52  A0A395Z784  MSIMHKKLMQIVLVKGLDIQDIVTLKKYDKKVKKFLTNKIGDDILI...  Negative   \n",
       "53  A0A349XYA6  MSGFTTVLALIVMGLVVIVLIRGLFNMLKGQDANRSNKLMQLRLLL...  Negative   \n",
       "54  A0A1M6C5T9  MILRGIALSLGKVLAVLAALAALASRFTSCLLCVYQPEMPQKVRNM...  Negative   \n",
       "55  A0A6I5N5G3  MEDFFRNVSRYPRYFISFTLGVLYNALQPLLPLLKRPTTAVALIGS...  Negative   \n",
       "56  A0A1H7WRT9  MKKLSNLEGVVVLSKKAQKTVHGGDYVMSSCSDYCPGTVQGNIFTG...  Negative   \n",
       "57  A0A7W0SMF1  MTDPRNDDKKDGSNWVGGSVLGVGIGAGVGAIVGNLGNGIWIGAVI...  Negative   \n",
       "58      B4Y7M7             TDSSIISAGWEPTEENEDERDRHPSYSGSGIDDDEDFI  Negative   \n",
       "59  A0A1C4D5Q1       MGRTKQGNRNAQANTNAKRGNRTSSEMVEFTTGRAKAEKNRPQD  Negative   \n",
       "60  A0A2H6JUL2               MLIGEAEFKLRAVKILFENGEYSDVISRALSYLVIS  Negative   \n",
       "61  A0A379M3S6  MGLRGSGSAARTHPLSAAPNFFTSGLHAAFDDGMQAVEWNFAGQGP...  Negative   \n",
       "62      B0N079  MKRIYRSRQDRMVCGVCGGIAEYFDLDPSLVRLGWIIFSAMGGSGF...  Negative   \n",
       "63  A0A1M6PH39  MNLDTFQNKLILILSYVDKLKRENVPINTQRILIQTYANDLEINLT...  Negative   \n",
       "64  A0A0A0NJ57  MREPCISDRDRGWFKSSYSGAGNTECVEAAFRPGSTAVRDSKDPGG...  Negative   \n",
       "65  A0A7Z2XC48  MNLSSMVKGILLALIFLGLMAGVLPSGAQLVRMGGSSWFLVGQSAE...  Negative   \n",
       "66  A0A1V3WIU2  MHRPIQPDSQISTGLTSRSGNAGDTAADAELRIGPQRSAHRLSAVV...  Negative   \n",
       "67  A0A1M6SHT4  MKRTFQPHNRKRVNKHGFMTRMEDRWGRAVLSRRRAKGRKRLTISD...  Negative   \n",
       "\n",
       "    KI (nM) Predicted  \n",
       "0                 NaN  \n",
       "1                 NaN  \n",
       "2                 NaN  \n",
       "3                 NaN  \n",
       "4                 NaN  \n",
       "5                 NaN  \n",
       "6                 NaN  \n",
       "7                 NaN  \n",
       "8                 NaN  \n",
       "9                 NaN  \n",
       "10                NaN  \n",
       "11           3.953144  \n",
       "12                NaN  \n",
       "13                NaN  \n",
       "14                NaN  \n",
       "15                NaN  \n",
       "16                NaN  \n",
       "17                NaN  \n",
       "18                NaN  \n",
       "19                NaN  \n",
       "20                NaN  \n",
       "21                NaN  \n",
       "22                NaN  \n",
       "23                NaN  \n",
       "24                NaN  \n",
       "25                NaN  \n",
       "26                NaN  \n",
       "27                NaN  \n",
       "28                NaN  \n",
       "29                NaN  \n",
       "30                NaN  \n",
       "31                NaN  \n",
       "32                NaN  \n",
       "33                NaN  \n",
       "34                NaN  \n",
       "35                NaN  \n",
       "36                NaN  \n",
       "37                NaN  \n",
       "38                NaN  \n",
       "39                NaN  \n",
       "40                NaN  \n",
       "41                NaN  \n",
       "42                NaN  \n",
       "43                NaN  \n",
       "44                NaN  \n",
       "45                NaN  \n",
       "46                NaN  \n",
       "47                NaN  \n",
       "48                NaN  \n",
       "49                NaN  \n",
       "50                NaN  \n",
       "51                NaN  \n",
       "52                NaN  \n",
       "53                NaN  \n",
       "54                NaN  \n",
       "55                NaN  \n",
       "56                NaN  \n",
       "57                NaN  \n",
       "58                NaN  \n",
       "59                NaN  \n",
       "60                NaN  \n",
       "61                NaN  \n",
       "62                NaN  \n",
       "63                NaN  \n",
       "64                NaN  \n",
       "65                NaN  \n",
       "66                NaN  \n",
       "67                NaN  "
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_rows',None)\n",
    "inferenceFasta('longhits.fasta', ensemble=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
